### ISSUE TITLE
Use Allen Features to optimize

### ISSUE THREAD

**Issue by russelljjarvis** [scidash/neuronunit#227]
This is mostly done there are about 40 features including waveform shape and spike timing. Optimizing would need to employ a sweep based approach to exploring current injections to find models with low error. Alternatively can optimize a cell relative to just one sweep, but have rheobase multiple as a gene.

**Comment by rgerkin** [comment]
I think behavior may be different in different sweeps, so probably better to use multiple sweeps.

**Comment by russelljjarvis** [comment]
I build an observation from one of many possible Allen data sweeps. On writing this, I just realized that the more efficient approach is to use compare model outputs to every sweep in an observation at first, and then on future model iterations to focus on the subset of sweeps that cause the least error. If I don't compare against all sweeps at first, then I am forced to make an arbitrary choice that will result in bias.

**Comment by russelljjarvis** [comment]
I have thought of a good solution: Reappropriate the Rheobase algorithm, so search for any target spike number. Just make sure that the model output has the same spike count as the sweep you are optimizing against.

**Comment by rgerkin** [comment]
That is clever.  But it also assumes that spike number is the correct criterion for to match on.  The alternative is for the the current injection amplitude to be the criterion to match on.  If at X pA, the model has 3 spikes and the data has 8 spike, the input resistance of the model can always be lowered (through some model parameter) to get to ~8 spikes.  Shouldn't the optimizer be able to find this automatically?

**Comment by russelljjarvis** [comment]
I agree, that was my first thought too, but I was finding using current injection values model versus experiment lead to bad matches on its own. What about using both?

**Comment by russelljjarvis** [comment]
Yes the optimizer could find spike number almost automatically if I make spike count a gene to be optimized against. That is actually a good idea and involves less re-writing of code: Derive an error from a spike count.

**Comment by rgerkin** [comment]
Perfect.  Make the error measure for spike count something like 
`error = logistic(abs(sqrt(n_spikes_data) - sqrt(n_spikes_model)))`, which will range from 0 (best, no difference) to 1 (infinite difference).  The sqrt stabilizes the variance across a range of spike counts.  For example, getting 1 spike when 0 is expected is a bigger difference than 21 spikes when 20 were expected, so a simple difference isn't a good measure, but neither is a ratio (1/0 is infinity).  The difference of square roots is a good balance and used in some other applications.  The logistic(abs(x)) then changes the range from (-inf, inf) to (0,1).  

These can later become sciunit score types but for now you can just do the whole calculation in `compute_score` and use a FloatScore or something like that.

**Comment by russelljjarvis** [comment]
Okay, will do that now.

**Comment by russelljjarvis** [comment]
There is also a question of how to assign mean and standard deviation to the allen sweep data observations. To prototype I set standard deviation to 1.0. I don't know if each sweep is repeated per cell, with only n=1, I am not sure what else I can do.

**Comment by rgerkin** [comment]
If you are fitting a model for that cell in particular (and not a mean across cells), and there is only one sweep with those parameters (e.g. current injection amplitude) then you probably shouldn't use a standard deviation at all, so instead of a ZScore you can use a RatioScore or something else.  If each sweep is repeated you can use the standard deviation across sweeps with ZScore.  If you insist on using ZScore even with one sweep a better "placeholder" standard deviation might be the sqrt(value) for positively valued things, or sqrt(value + offset) if value can be zero, such as for spike count, where offset is something like 1.

**Comment by russelljjarvis** [comment]
I recently tried using RatioScore I believe it throws errors at extreme values. ZScore seems to be more robust at around extreme values, thus I keep returning to ZScore.
